packages/salesforcedx-vscode-core/test/vscode-integration/commands/forceOpenDocumentation.test.ts

```diff
@@ -0,0 +1,125 @@
+  auraDocUrl,
+  defaultDocUrl,
+  forceOpenDocumentation,
+  functionsDocUrl,
+  lwcDocUrl,
+  soqlDocUrl
+} from '../../../src/commands';
+
+describe('forceOpenDocumentation', () => {
+  let sb: SinonSandbox;
+  let activeTextEditorStub: SinonStub;
+  let openExternalStub: SinonStub;
+
+  beforeEach(() => {
+    sb = createSandbox();
+    activeTextEditorStub = sb.stub(vscode.window, 'activeTextEditor');
+    openExternalStub = sb.stub(vscode.env, 'openExternal');
+  });
+
+  afterEach(() => {
+    sb.restore();
+  });
+
+  it('should open the documentation for Aura', async () => {
+    activeTextEditorStub.get(() => ({
+      document: {
+        fileName:
+          '/force-app/main/default/aura/exampleAuraComponent/exampleAuraComponent.cmp'
+      }
+    }));
+
+    await forceOpenDocumentation();
+
+    expect(openExternalStub.getCall(0).args[0].toString()).to.equal(auraDocUrl);
+  });
+
+  it('should open the documentation for Apex class', async () => {
+    activeTextEditorStub.get(() => ({
+      document: {
+        fileName: '/force-app/main/default/classes/exampleApexClass.cls'
+      }
+    }));
+
+    await forceOpenDocumentation();
+
+    expect(openExternalStub.getCall(0).args[0].toString()).to.equal(apexDocUrl);
+  });
+
+  it('should open the documentation for Anonymous Apex', async () => {
+    activeTextEditorStub.get(() => ({
+      document: {
+        fileName: '/scripts/apex/exampleApex.apex'
+      }
+    }));
+
+    await forceOpenDocumentation();
+
+    expect(openExternalStub.getCall(0).args[0].toString()).to.equal(apexDocUrl);
+  });
+
+  it('should open the documentation for SOQL', async () => {
+    activeTextEditorStub.get(() => ({
+      document: {
+        fileName: '/scripts/soql/exampleSoql.soql'
+      }
+    }));
+
+    await forceOpenDocumentation();
+
+    expect(openExternalStub.getCall(0).args[0].toString()).to.equal(soqlDocUrl);
+  });
+
+  it('should open the documentation for LWC', async () => {
+    activeTextEditorStub.get(() => ({
+      document: {
+        fileName:
+          '/force-app/main/default/lwc/exampleLwcComponent/exampleLwcComponent.js'
+      }
+    }));
+
+    await forceOpenDocumentation();
+
+    expect(openExternalStub.getCall(0).args[0].toString()).to.equal(lwcDocUrl);
+  });
+
+  it('should open the documentation for Functions', async () => {
+    activeTextEditorStub.get(() => ({
+      document: {
+        fileName: '/functions/example/index.js'
+      }
+    }));
+
+    await forceOpenDocumentation();
+
+    expect(openExternalStub.getCall(0).args[0].toString()).to.equal(
+      functionsDocUrl
+    );
+  });
+
+  it('should open the default documentation', async () => {
+    activeTextEditorStub.get(() => ({
+      document: {
+        fileName: '/force-app/main/default/staticresources/example-image.png'
+      }
+    }));
+
+    await forceOpenDocumentation();
+
+    expect(openExternalStub.getCall(0).args[0].toString()).to.equal(
+      defaultDocUrl
+    );
+  });
```

Comment:
nice tests.  Are there any failure conditions we can verify? What happens if `vscode.env.openExternal` throws?

Reply:
I wasn't able to think of and failure conditions, and I haven't found cases where openExternal fails.

Reply:
It's a corner case, but what if the user doesn't have a default browser? It's usually good to know the behaviour (via a unit test) even if we don't expect failure. It documents what we expect to happen.  Like if we can't open a browser we should probably be outputting to the user that something went wrong verse just silently doing nothing.

Reply:
This is something we get automatically from VSC, and I think writing a unit test for it is writing a test validating VS Code, and doesn't validate our code.  As an example, I ran this:
```
export async function forceOpenDocumentation() {
  let docUrl = '';
  const editor = vscode.window.activeTextEditor;
  if (editor) {
    const filePath = editor.document.fileName;
    const extension = path.extname(filePath);

    throw Error('OK dont do that');
.
.
.
```

And the user sees this:
![image](https://user-images.githubusercontent.com/46358907/189424719-6a41262b-9934-45c1-9aa3-0817fc0d54f1.png)

Reply:
cool.  The main thing I'd have expected in the unit test would be validating that the error thrown by the called function is what is get thrown from our calling code.  Granted there's not error handling here so we don't have anything extra to verify so I'm cool with just calling it a day. Error conditions is just always something to consider in unit tests.